be/src/storage/tablet_reader.cpp

```diff
@@ -95,12 +98,124 @@ Status TabletReader::open(const TabletReaderParams& read_params) {
+    if (rowids.size() != 1) {
+        return Status::InternalError(strings::Substitute("get rowid size not match tablet:$0 $1 != $2",
+                                                         _tablet->tablet_id(), rowids.size(), 1));
+    }
+    // do not check read version in use_pk_index mode
+    uint32_t rssid = rowids[0] >> 32;
+    uint32_t rowid = rowids[0] & 0xffffffff;
+    if (rssid == (uint32_t)-1) {
+        _collect_iter = new_empty_iterator(_schema, _reader_params->chunk_size);
+        return Status::OK();
+    }
+
+    RowsetSharedPtr rowset;
+    uint32_t segment_idx = 0;
+    RETURN_IF_ERROR(_tablet->updates()->get_rowset_and_segment_idx_by_rssid(rssid, &rowset, &segment_idx));
+
+    RowsetReadOptions rs_opts;
+    rs_opts.predicates = pushdown_predicates;
+    rs_opts.sorted = false;
+    rs_opts.reader_type = _reader_params->reader_type;
+    rs_opts.chunk_size = _reader_params->chunk_size;
+    rs_opts.delete_predicates = &_delete_predicates;
+    rs_opts.stats = &_stats;
+    rs_opts.runtime_state = _reader_params->runtime_state;
+    rs_opts.profile = _reader_params->profile;
+    rs_opts.use_page_cache = _reader_params->use_page_cache;
+    rs_opts.tablet_schema = &_tablet->tablet_schema();
+    rs_opts.global_dictmaps = _reader_params->global_dictmaps;
+    rs_opts.unused_output_column_ids = _reader_params->unused_output_column_ids;
+    rs_opts.runtime_range_pruner = _reader_params->runtime_range_pruner;
+    // single row fetch, no need to use delvec
+    rs_opts.is_primary_keys = false;
+
+    rs_opts.rowid_range_option = std::make_shared<RowidRangeOption>();
+    auto rowid_range = std::make_shared<SparseRange>();
+    rowid_range->add({rowid, rowid + 1});
+    if (segment_idx >= rowset->num_segments()) {
+        return Status::InternalError(strings::Substitute("segment_idx out of range tablet:$0 $1 >= $2",
+                                                         _tablet->tablet_id(), segment_idx, rowset->num_segments()));
+    }
+    rs_opts.rowid_range_option->add(rowset.get(), rowset->segments()[segment_idx].get(), rowid_range);
+
+    std::vector<ChunkIteratorPtr> iters;
+    RETURN_IF_ERROR(rowset->get_segment_iterators(schema(), rs_opts, &iters));
+
+    if (iters.size() != 1) {
+        return Status::InternalError(
+                strings::Substitute("get_segment_iterators for pointer query should return single iter tablet:$0 $1",
+                                    _tablet->tablet_id(), iters.size()));
+    }
+
+    _collect_iter = iters[0];
+
+    // other collector setup
+    RETURN_IF_ERROR(_collect_iter->init_encoded_schema(*_reader_params->global_dictmaps));
+    RETURN_IF_ERROR(_collect_iter->init_output_schema(*_reader_params->unused_output_column_ids));
+
+    return Status::OK();
+}
+
 Status TabletReader::do_get_next(Chunk* chunk) {
     DCHECK(!_is_vertical_merge);
+    if (_collect_iter == nullptr) {
```

Comment:
Why not place the `_init_collector_for_pk_index_read` in the open() function?

Reply:
open() is not in pipeline IO thread, this index read will read index file.